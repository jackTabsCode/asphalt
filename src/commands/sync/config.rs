use anyhow::Context;
use serde::{Deserialize, Serialize};
use std::{
    collections::HashMap,
    fmt::{self, Display, Formatter},
};
use tokio::fs::{read_to_string, write};

#[derive(Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum CodegenStyle {
    Flat,
    Nested,
}

impl Display for CodegenStyle {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            CodegenStyle::Flat => write!(f, "Flat"),
            CodegenStyle::Nested => write!(f, "Nested"),
        }
    }
}

#[derive(Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum CreatorType {
    User,
    Group,
}

impl Display for CreatorType {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            CreatorType::User => write!(f, "User"),
            CreatorType::Group => write!(f, "Group"),
        }
    }
}

#[derive(Debug, Deserialize, Serialize)]
pub struct Creator {
    #[serde(rename = "type")]
    pub creator_type: CreatorType,
    pub id: u64,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct ExistingAsset {
    pub id: u64,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct CodegenConfig {
    pub output_name: Option<String>,
    pub typescript: Option<bool>,
    pub style: Option<CodegenStyle>,
    pub strip_extension: Option<bool>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct SyncConfig {
    pub asset_dir: String,
    #[serde(default)]
    pub exclude_assets: Vec<String>,
    pub write_dir: String,
    pub creator: Creator,
    pub codegen: CodegenConfig,
    pub existing: Option<HashMap<String, ExistingAsset>>,
}

static FILE_NAME: &str = "asphalt.toml";

impl SyncConfig {
    pub async fn read() -> anyhow::Result<Self> {
        let content = read_to_string(FILE_NAME)
            .await
            .context("Failed to read config. Did you create it?")?;
        toml::from_str(&content).context("Failed to parse config")
    }

    pub async fn write(&self) -> anyhow::Result<()> {
        let content = toml::to_string(self)?;
        write(FILE_NAME, content)
            .await
            .context("Failed to write config")
    }
}
