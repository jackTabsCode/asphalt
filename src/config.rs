use std::collections::HashMap;

use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum CreatorType {
    User,
    Group,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct Creator {
    #[serde(rename = "type")]
    pub creator_type: CreatorType,
    pub id: u64,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct ExistingAsset {
    pub id: u64,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct Config {
    pub asset_dir: String,
    pub write_dir: String,
    pub creator: Creator,
    pub output_name: Option<String>,
    pub typescript: Option<bool>,
    pub luau: Option<bool>,

    pub existing: Option<HashMap<String, ExistingAsset>>,
}
