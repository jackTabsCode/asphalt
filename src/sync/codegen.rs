use crate::config::{Codegen, CodegenStyle};
use anyhow::bail;
use std::{
    collections::BTreeMap,
    path::{Path, PathBuf},
};

pub type CodegenInput = BTreeMap<PathBuf, String>;

pub enum CodegenNode {
    Table(BTreeMap<String, CodegenNode>),
    String(String),
    #[allow(dead_code)]
    Number(u64),
}

pub enum CodegenLanguage {
    TypeScript,
    Luau,
}

pub fn from_codegen_input(input: &CodegenInput, config: &Codegen) -> CodegenNode {
    let mut root = CodegenNode::Table(BTreeMap::new());

    for (path, value) in input {
        match config.style {
            CodegenStyle::Nested => {
                let components = normalize_path_components(path, config.strip_extensions);
                insert_nested(&mut root, &components, value);
            }
            CodegenStyle::Flat => {
                let key = normalize_path_string(path, config.strip_extensions);
                insert_flat(&mut root, &key, value);
            }
        }
    }

    root
}

fn normalize_path_components(path: &Path, strip_extensions: bool) -> Vec<String> {
    let mut components: Vec<String> = Vec::new();
    let total_components = path.iter().count();

    for (i, comp) in path.iter().enumerate() {
        if i == total_components - 1 && strip_extensions {
            let as_path = Path::new(comp);
            if let Some(stem) = as_path.file_stem() {
                components.push(stem.to_string_lossy().to_string());
                continue;
            }
        }
        components.push(comp.to_string_lossy().to_string());
    }
    components
}

fn normalize_path_string(path: &Path, strip_extensions: bool) -> String {
    if strip_extensions {
        if let (Some(file_name), Some(parent)) = (path.file_name(), path.parent()) {
            if let Some(stem) = Path::new(file_name).file_stem() {
                let parent_str = parent.to_string_lossy();
                return if parent_str.is_empty() || parent_str == "." {
                    stem.to_string_lossy().into_owned()
                } else {
                    format!("{}/{}", parent_str, stem.to_string_lossy())
                };
            }
        }
    }
    path.to_string_lossy().into_owned()
}

fn insert_flat(node: &mut CodegenNode, key: &str, content: &str) {
    match node {
        CodegenNode::Table(map) => {
            map.insert(key.into(), CodegenNode::String(content.into()));
        }
        _ => {
            *node = CodegenNode::Table(BTreeMap::new());
            if let CodegenNode::Table(map) = node {
                map.insert(key.into(), CodegenNode::String(content.into()));
            }
        }
    }
}

fn insert_nested(node: &mut CodegenNode, components: &[String], content: &str) {
    if !matches!(node, CodegenNode::Table(_)) {
        *node = CodegenNode::Table(BTreeMap::new());
    }

    if components.is_empty() {
        return;
    }

    if let CodegenNode::Table(map) = node {
        let component = &components[0];

        if components.len() == 1 {
            map.insert(component.clone(), CodegenNode::String(content.into()));
        } else {
            let next_node = map
                .entry(component.clone())
                .or_insert_with(|| CodegenNode::Table(BTreeMap::new()));

            if !matches!(next_node, CodegenNode::Table(_)) {
                *next_node = CodegenNode::Table(BTreeMap::new());
            }

            insert_nested(next_node, &components[1..], content);
        }
    }
}

pub fn generate_code(
    lang: CodegenLanguage,
    name: &str,
    node: &CodegenNode,
) -> anyhow::Result<String> {
    if !matches!(node, CodegenNode::Table(_)) {
        bail!("Root node must be a Table");
    }

    Ok(match lang {
        CodegenLanguage::TypeScript => generate_typescript(name, node),
        CodegenLanguage::Luau => generate_luau(name, node),
    })
}

fn generate_typescript(name: &str, node: &CodegenNode) -> String {
    let body = generate_ts_node(node, 0);
    format!("declare const {}: {}\n\nexport = {}", name, body, name)
}

fn generate_ts_node(node: &CodegenNode, indent: usize) -> String {
    match node {
        CodegenNode::Table(map) => {
            let mut result = String::from("{\n");
            for (k, v) in map {
                result.push_str(&"\t".repeat(indent + 1));
                let k = if is_valid_identifier(k) {
                    k.clone()
                } else {
                    format!("\"{}\"", k)
                };
                result.push_str(&k);
                result.push_str(": ");
                result.push_str(&generate_ts_node(v, indent + 1));
                result.push('\n');
            }
            result.push_str(&"\t".repeat(indent));
            result.push('}');
            result
        }
        CodegenNode::String(_) => "string".to_string(),
        CodegenNode::Number(_) => "number".to_string(),
    }
}

fn generate_luau(name: &str, node: &CodegenNode) -> String {
    let body = generate_luau_node(node, 0);
    format!("local {} = {}\n\nreturn {}", name, body, name)
}

fn generate_luau_node(node: &CodegenNode, indent: usize) -> String {
    match node {
        CodegenNode::Table(map) => {
            let mut result = String::from("{\n");
            for (k, v) in map {
                result.push_str(&"\t".repeat(indent + 1));
                let k = if is_valid_identifier(k) {
                    k.clone()
                } else {
                    format!("[\"{}\"]", k)
                };
                result.push_str(&k);
                result.push_str(" = ");
                result.push_str(&generate_luau_node(v, indent + 1));
                result.push_str(",\n");
            }
            result.push_str(&"\t".repeat(indent));
            result.push('}');
            result
        }
        CodegenNode::String(s) => format!("\"{}\"", s),
        CodegenNode::Number(n) => format!("{}", n),
    }
}

fn is_valid_ident_char_start(value: char) -> bool {
    value.is_ascii_alphabetic() || value == '_'
}

fn is_valid_ident_char(value: char) -> bool {
    value.is_ascii_alphanumeric() || value == '_'
}

fn is_valid_identifier(value: &str) -> bool {
    let mut chars = value.chars();

    match chars.next() {
        Some(first) => {
            if !is_valid_ident_char_start(first) {
                return false;
            }
        }
        None => return false,
    }

    chars.all(is_valid_ident_char)
}
