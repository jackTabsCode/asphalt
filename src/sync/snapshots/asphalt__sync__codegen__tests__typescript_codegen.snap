---
source: src/sync/codegen.rs
expression: code
---
// This file is automatically @generated by Asphalt.
// It is not intended for manual editing.
declare const name: {
	fred: string
	qux: {
		baz: number
		foo: string
	}
	waldo: Content
}

export = name
